### ISSUE TITLE
question: "The receiver is overrun by more signals than expected (bounded queue...)" in 1.3.0-M1

### ISSUE THREAD

**Issue by tchupp** [reactor/reactor-kafka#162]
(I couldn't find anything on Stack Overflow referencing this)

After upgrading from `1.2.2.RELEASE` to `1.3.0-M1`, we now see "The receiver is overrun by more signals than expected (bounded queue...)" in a few (not all) of our flows that read from Kafka.

Full StackTrace:
```
reactor.core.Exceptions$OverflowException: The receiver is overrun by more signals than expected (bounded queue...)
	at reactor.core.Exceptions.failWithOverflow(Exceptions.java:221)
	at reactor.core.publisher.FluxCreate$ErrorAsyncSink.onOverflow(FluxCreate.java:687)
	at reactor.core.publisher.FluxCreate$NoOverflowBaseAsyncSink.next(FluxCreate.java:652)
	at reactor.core.publisher.FluxCreate$SerializedSink.next(FluxCreate.java:153)
	at reactor.kafka.receiver.internals.ConsumerEventLoop$PollEvent.run(ConsumerEventLoop.java:211)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

I dug in to the differences between the "poll" implementations in `1.2.2.RELEASE` and `1.3.0-M1` and noticed that there are a few differences that _might_ cause this..

1. Changing back pressure behavior from "Ignore" to "Error": This seems like the most likely candidate for causing the back pressure exception downstream, but by itself does not explain why back pressure occurred.
2. Changing the buffer size of the "ConsumerRecords" processor from 256 to 1: Also a likely candidate for causing the back pressure exception.

Has anyone else encountered this?
Can this be solved by adding a `.onBackpressureBuffer()` after a `KafkaReceiver.receive()`?
If the polls to Kafka only happen when signals are requested, how is this happening?

[Full diff](https://github.com/reactor/reactor-kafka/compare/v1.2.2.RELEASE...v1.3.0-M1#)
Relevant code blocks from each version:
`1.2.2.RELEASE`
```
private synchronized Flux<ConsumerRecords<K, V>> createConsumerFlux() {
...
    recordEmitter = EmitterProcessor.create(); <- default buffer size is 256
    recordSubmission = recordEmitter.sink(); <- default behavior is IGNORE
...
}
...
    private class PollEvent extends Event<ConsumerRecords<K, V>> {	

        private AtomicInteger pendingCount = new AtomicInteger();	
        private final Duration pollTimeout;	
        private final AtomicBoolean partitionsPaused = new AtomicBoolean();	
        PollEvent() {	
            super(EventType.POLL);	
            pollTimeout = receiverOptions.pollTimeout();	
        }	
        @Override	
        public void run() {	
            needsHeartbeat.set(false);	
            try {	
                if (isActive.get()) {	
                    // Ensure that commits are not queued behind polls since number of poll events is	
                    // chosen by reactor.	
                    commitEvent.runIfRequired(false);	
                    pendingCount.decrementAndGet();	
                    if (requestsPending.get() > 0 && !awaitingTransaction.get()) {	
                        if (partitionsPaused.getAndSet(false))	
                            consumer.resume(consumer.assignment());	
                    } else {	
                        if (!partitionsPaused.getAndSet(true))	
                            consumer.pause(consumer.assignment());	
                    }	

                    ConsumerRecords<K, V> records = consumer.poll(pollTimeout);	
                    if (records.count() > 0) {	
                        recordSubmission.next(records);	
                    }	
                    if (isActive.get()) {	
                        int count = ((ackMode == AckMode.AUTO_ACK || ackMode == AckMode.EXACTLY_ONCE) && records.count() > 0) ? 1 : records.count();	
                        if (requestsPending.get() == Long.MAX_VALUE || requestsPending.addAndGet(0 - count) > 0 || commitEvent.inProgress.get() > 0)	
                            scheduleIfRequired();	
                    }	
                }	
            } catch (Exception e) {	
                if (isActive.get()) {	
                    log.error("Unexpected exception", e);	
                    fail(e);	
                }	
            }	
        }	

        void scheduleIfRequired() {	
            if (pendingCount.get() <= 0) {	
                emit(this);	
                pendingCount.incrementAndGet();	
            }	
        }	
    }	
```

`1.3.0-M1`
```
class ConsumerHandler<K, V> {
...
    private final FluxIdentityProcessor<ConsumerRecords<K, V>> processor = Processors.more().multicast(1); <- max buffer size of 1
...
        consumerEventLoop = new ConsumerEventLoop<>(
            ackMode,
            atmostOnceOffsets,
            receiverOptions,
            eventScheduler,
            consumer,
            isRetriableException,
            processor.sink(FluxSink.OverflowStrategy.ERROR), <- this is the "sink" referenced below
            awaitingTransaction
        );
...
    class PollEvent implements Runnable {

        private final Duration pollTimeout = receiverOptions.pollTimeout();

        @Override
        public void run() {
            try {
                if (isActive.get()) {
                    // Ensure that commits are not queued behind polls since number of poll events is
                    // chosen by reactor.
                    commitEvent.runIfRequired(false);
                    if (sink.requestedFromDownstream() > 0) {
                        if (!awaitingTransaction.get()) {
                            consumer.resume(consumer.assignment());
                        } else {
                            consumer.pause(consumer.assignment());
                            schedule();
                        }
                    } else {
                        consumer.pause(consumer.assignment());
                    }

                    ConsumerRecords<K, V> records = consumer.poll(pollTimeout);
                    if (isActive.get()) {
                        if (sink.requestedFromDownstream() > 1 || commitEvent.inProgress.get() > 0) {
                            schedule();
                        }
                    }
                    sink.next(records); <- no longer checks if "records" contain anything before submitting
                }
            } catch (Exception e) {
                if (isActive.get()) {
                    log.error("Unexpected exception", e);
                    sink.error(e);
                }
            }
        }

        void schedule() {
            eventScheduler.schedule(this);
        }
    }
```

Reactor version(s) used:
- reactor-kafka `1.3.0-M1`
- reactor-core `3.3.4.RELEASE`

**Comment by bsideup** [comment]
@tchupp if it happens in most of your consumers, could you please try to create a reproducer that we can run and debug? It will help a lot!

**Comment by tchupp** [comment]
It only happens in some of our consumers; we are having trouble reproducing this consistently 😅

**Comment by tchupp** [comment]
We have still not been able to consistently reproduce this issue; but in every occurrence we have seen so far, this exception is preceded by this log:
```
[Consumer clientId=<client_id>, groupId=<group_id>] Member <member_id> sending LeaveGroup request to coordinator <node>
```
We do _not_ see the WARN log containing "This member will leave the group because consumer poll timeout has expired....", leading us to believe the "LeaveGroup" may have been triggered by a rebalance.

We are using `kafka-clients:2.3.1`, so we don't get the nice "due to {}" that we would see with `kafka-clients:2.4.1+`

**Comment by tchupp** [comment]
This commit looks well positioned to address the root error we were seeing (receiver overrun by more signals than expected): https://github.com/reactor/reactor-kafka/commit/3150a5a9413b3bd4ade0a9354d81cc22e1ea3d27

Specifically, the change from
```
sink.next(records);
```
to
```
Operators.produced(REQUESTED, ConsumerEventLoop.this, 1);
while (sink.tryEmitNext(records) == Sinks.Emission.FAIL_OVERFLOW && isActive.get()) {
    LockSupport.parkNanos(10);
}
```
in the `PollEvent` class.

**Comment by bsideup** [comment]
@tchupp there will be another change soon that should improve the backpressure handling even more.

Meanwhile, if you can give the SNAPSHOTs a try and report, it will be highly appreciated!

**Comment by bsideup** [comment]
Hi @tchupp,

We just released `1.3.0.RELEASE`, please let us know if the issue remains. Thanks!

**Comment by tchupp** [comment]
Will do! Thank you
